mapper/api_views.py

```diff
@@ -0,0 +1,141 @@
+    """
+    Receives optouts from RapidPro and reverses the migration for the specified
+    user.
+    """
+    authentication_classes = (TokenAuthentication,)
+    rapidpro_client = TembaClient(
+        settings.RAPIDPRO_URL, settings.RAPIDPRO_TOKEN)
+    sbm_client = StageBasedMessagingApiClient(
+        settings.STAGE_BASED_MESSAGING_TOKEN,
+        settings.STAGE_BASED_MESSAGING_URL)
+
+    def __init__(self, *args, **kwargs):
+        self.messagesets = {}
+        super(RapidproOptout, self).__init__(*args, **kwargs)
+
+    def get_rapidpro_contact(self, uuid):
+        """
+        Retrieves the full data for a rapidpro contact, given the UUID.
+        """
+        contact = self.rapidpro_client.get_contacts(uuid).first()
+        if contact is None:
+            raise NotFound('Rapidpro contact {} does not exist'.format(uuid))
+        return contact
+
+    def find_migration(self, identity_uuid):
+        """
+        Finds the latest MigrateSubscription for the given identity UUID.
+        """
+        try:
+            migration = MigratedIdentity.objects\
+                .filter(identity_uuid=identity_uuid)\
+                .select_related('migrate_subscription')\
+                .latest('migrate_subscription__created_at')\
+                .migrate_subscription
+        except MigratedIdentity.DoesNotExist:
+            raise InvalidRapidproContact(
+                'Seed identity {} does not have any migrations.'.format(
+                    identity_uuid))
+        return migration
+
+    def get_messageset(self, messageset_id):
+        if messageset_id not in self.messagesets:
+            self.messagesets[messageset_id] = self.sbm_client.get_messageset(
+                messageset_id)
+        return self.messagesets[messageset_id]
+
+    def get_existing_subscriptions(self, identity_uuid, messageset):
+        """
+        Gets all active subscriptions to the specified messageset for the
+        specified identity.
+        """
+        return list(self.sbm_client.get_subscriptions({
+            'identity': identity_uuid,
+            'messageset': messageset,
+            'active': True,
+        })['results'])
+
+    def create(self, request):
+        """
+        Reverses the migration for the specified user.
+        """
+        serializer = RapidproOptoutSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        contact = self.get_rapidpro_contact(
+            serializer.validated_data['contact'])
+
+        try:
+            seed_uuid = UUID(contact.fields[settings.RAPIDPRO_UUID_FIELD])
+        except (ValueError, TypeError, KeyError):
+            raise InvalidRapidproContact(
+                'Rapidpro contact {} has an invalid {} field.'.format(
+                    contact.uuid, settings.RAPIDPRO_UUID_FIELD))
+
+        migration = self.find_migration(seed_uuid)
+        existing_subs = self.get_existing_subscriptions(
+            seed_uuid, migration.to_messageset)
```

Comment:
what if the gates subscriptions have multiple messagesets? eg a `gates_prebirth` that has a `gates_postbirth` as the `next_messageset`? Then the migration may point to the first messageset but the user may have moved onto the second messageset.
I _think_ this is a possible design for some of the investigations

Reply:
Yup, this is something I hadn't thought of. Let me chat to some people and get back to you on whether this is a valid use case.